### ISSUE TITLE
arm64: building CONFIG_KVM with LTO fails due to dynamic patching in inline assembly

### ISSUE THREAD

**Issue by samitolvanen** [ClangBuiltLinux/linux#472]
Compiling the arm64 defconfig with LTO, I noticed LLVM's integrated assembler doesn't understand ```ALTERNATIVE_CB```:
```
<inline asm>:12:6: error: expected absolute expression
 .if kvm_update_va_mask == 0
     ^
<inline asm>:21:6: error: expected absolute expression
 .if kvm_update_va_mask == 0
     ^
<inline asm>:24:2: error: unrecognized instruction mnemonic
        NOT_AN_INSTRUCTION
        ^
LLVM ERROR: Error parsing inline asm
```

The inline assembly is in arch/arm64/include/asm/kvm_mmu.h:
```C
static inline unsigned long __kern_hyp_va(unsigned long v)
{
        asm volatile(ALTERNATIVE_CB("and %0, %0, #1\n"
                                    "ror %0, %0, #1\n"
                                    "add %0, %0, #0\n"
                                    "add %0, %0, #0, lsl 12\n"
                                    "ror %0, %0, #63\n",
                                    kvm_update_va_mask)
                     : "+r" (v));
        return v;
}
```

And the definition for ```ALTERNATIVE_CB``` in arch/arm64/include/asm/alternative.h:
```C
#define __ALTERNATIVE_CFG(oldinstr, newinstr, feature, cfg_enabled, cb) \
        ".if "__stringify(cfg_enabled)" == 1\n"                         \
        "661:\n\t"                                                      \
        oldinstr "\n"                                                   \
        "662:\n"                                                        \
        ".pushsection .altinstructions,\"a\"\n"                         \
        ALTINSTR_ENTRY(feature,cb)                                      \
        ".popsection\n"                                                 \
        " .if " __stringify(cb) " == 0\n"                               \
        ".pushsection .altinstr_replacement, \"a\"\n"                   \
        "663:\n\t"                                                      \
        newinstr "\n"                                                   \
        "664:\n\t"                                                      \
        ".popsection\n\t"                                               \
        ".org   . - (664b-663b) + (662b-661b)\n\t"                      \
        ".org   . - (662b-661b) + (664b-663b)\n"                        \
        ".else\n\t"                                                     \
        "663:\n\t"                                                      \
        "664:\n\t"                                                      \
        ".endif\n"                                                      \
        ".endif\n"

#define ALTERNATIVE_CB(oldinstr, cb) \
        __ALTERNATIVE_CFG(oldinstr, "NOT_AN_INSTRUCTION", ARM64_CB_PATCH, 1, cb)
```

This code was introduced a couple of years ago in commit dea5e2a4c5bcf196f879a66cebdcca07793e8ba4.

**Comment by nickdesaulniers** [comment]
@samitolvanen , without LTO (but with `-no-integrated-as` removed), can you provide a preprocessed source of a translation unit that calls `__kern_hyp_va`?  You should be able to `make <clang args> <translation_unit.o> V=1` then switch the clang invocation to have `-E` instead of `-o <translation_unit.o>`.

Then it should be easier to narrow down what fails to parse, as we'll just have 1 big string of asm.

**Comment by nickdesaulniers** [comment]
oh, nvm, the error message does kind of explain it (though a pared down test case would make it more straightforward to understand and compare against GAS).

> error: expected absolute expression

Looks like `kvm_update_va_mask` should be a global symbol referring to a function, I believe.  Probably the first of what may be 2 bugs here.

> error: unrecognized instruction mnemonic NOT_AN_INSTRUCTION

I wonder if there's an equivalent of `ud2` on ARM64, or if there's some kind of peephole optimization that's removing this instruction (since the address of a global function should never be `== 0`)?

**Comment by nickdesaulniers** [comment]
Simple reproducer (I feel like I've discussed this before, maybe with @gwelymernans and/or @smithp35 ):
```gas
// clang -target aarch64-linux-gnu x.s -c
.global _start
.text
_start:
  .if _bar == 0
  mov w0, 42
  .else
  mov w0, 43
  .endif
  ret

_bar:
  ret
```

**Comment by nickdesaulniers** [comment]
yeah reminds me of #473 .

**Comment by tpimh** [comment]
I also hit this error when compiling with integrated assembler. Here is the error message for reference:
```
./arch/arm64/include/asm/kvm_mmu.h:97:15: error: expected absolute expression
        asm volatile(ALTERNATIVE_CB("and %0, %0, #1\n"
                     ^
./arch/arm64/include/asm/alternative.h:91:2: note: expanded from macro 'ALTERNATIVE_CB'
        __ALTERNATIVE_CFG(oldinstr, "NOT_AN_INSTRUCTION", ARM64_CB_PATCH, 1, cb)
        ^
./arch/arm64/include/asm/alternative.h:71:2: note: expanded from macro '__ALTERNATIVE_CFG'
        ALTINSTR_ENTRY(feature,cb)                                      \
        ^
./arch/arm64/include/asm/alternative.h:40:3: note: expanded from macro 'ALTINSTR_ENTRY'
        " .if " __stringify(cb) " == 0\n"                                     \
         ^
<inline asm>:12:6: note: instantiated into assembly here
 .if kvm_update_va_mask == 0
     ^
```

**Comment by nickdesaulniers** [comment]
@samitolvanen just submitted a patch for the kernel: https://lore.kernel.org/r/20191007211418.30321-1-samitolvanen@google.com/

**Comment by jcai19** [comment]
> Simple reproducer (I feel like I've discussed this before, maybe with @gwelymernans and/or @smithp35 ):
> 
> ```assembly
> // clang -target aarch64-linux-gnu x.s -c
> .global _start
> .text
> _start:
>   .if _bar == 0
>   mov w0, 42
>   .else
>   mov w0, 43
>   .endif
>   ret
> 
> _bar:
>   ret
> ```

There might be multiple issues associate with this error message. The direct cause of the failure in the above code was due to symbol content not set (SA.isUnset() and SB.isUnset() both be true in http://llvm.org/doxygen/MCExpr_8cpp_source.html#l00524). But I also found cases with the same error message when calculating distances between two labels belonging to two different sections respectively while building for arm, and failed to proceed because the layout was not finalized at that point (http://llvm.org/doxygen/MCExpr_8cpp_source.html#l00543). I need to further investigate what caused the first issue but for the second issue, I think we may need to implement another pass to replace dot with actual instruction address after layout is finalized.
